### ISSUE TITLE
AUROC Calculation

### ISSUE THREAD

**Issue by weijiacheng00** [icc2115/Neural-GC#13]
Dear iancovert

I hope this email finds you well.

I have been working with the code you provided on GitHub and had a few questions regarding the AUROC calculation. Specifically, I would like to understand how AUROC is computed by sweeping through a series of different 𝜆 regularization parameters.

Additionally, I am curious to know whether you use cross-validation to select the optimal λ and, if so, how this process is implemented. Lastly, when calculating AUROC, is the Granger Causality (GC) matrix continuous, or is it binarized for this computation?

I would greatly appreciate any clarification you can provide regarding these points.

Thank you very much for your time and assistance!

Best regards,

**Comment by weijiacheng00** [comment]
GC_est_flat = GC_est.ravel()
GC_flat = GC.ravel()


fpr, tpr, thresholds = roc_curve(GC_flat, GC_est_flat)
print(thresholds)

auroc_value = auc(fpr, tpr)
print(f'AUROC: {auroc_value:.4f}')  

Here is my method for calculating AUROC, where I set threshold=False when computing GC_est. Is this the correct approach?

**Comment by weijiacheng00** [comment]
This is the TPR and FPR calculated by scanning a series of values of λ on the Lorenz_96 simulation data, with T=250.
![image](https://github.com/user-attachments/assets/7c636eff-49e6-4693-bcee-e3b202faa87e)
Is this method correct?

**Comment by iancovert** [comment]
Hi, your understanding of our AUROC calculation sounds mostly correct. We trained models with a series of regularization strengths, extracted the GC matrix for each model (each one is a binary matrix), calculated the sensitivity/specificity by comparing the estimated and true GC matrices, plotted these points on a ROC curve, and finally calculated the area under the curve.

If I understand your code properly, it seems like you're probably not using the binarized matrix, and instead calculating the AUROC separately for each estimated non-thresholded GC matrix? I believe some papers do this, so it's not without precedent, and it's certainly faster to evaluate this way. However, the results are sensitive to the choice of regularization strength, which isn't ideal. Also, we didn't trust that we could reliably rank the non-sparsified features by their first-layer norms, so we preferred to simply binarize these values and re-train the model at different regularization strengths.

As for the question about selecting $\lambda$ via cross-validation: if you wanted to pick a single best regularization strength, I believe this is what you must do. Unfortunately, this model selection criterion could lead to overestimating the optimal $\lambda$ because having extra inputs will likely not hurt performance. We avoided this issue by not selecting a single $\lambda$ value, and instead tracing out the sensitivity/specificity for a range of values.

**Comment by weijiacheng00** [comment]
Can you tell me the λ range for each experiment? Thank you, I am unable to obtain the results in the paper. I suspect that my λ range was not appropriate.

**Comment by weijiacheng00** [comment]
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import auc

# Load the true Granger causality matrix and predicted matrix
GC_true = np.load("result/clstm_lorenz/250/GC.npy")

# List of file paths
file_paths = [
    "result/clstm_lorenz/250/gc_est_LX_0.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_0.10.npy",
    "result/clstm_lorenz/250/gc_est_LX_0.30.npy",
    "result/clstm_lorenz/250/gc_est_LX_0.50.npy",
    "result/clstm_lorenz/250/gc_est_LX_0.70.npy",
    "result/clstm_lorenz/250/gc_est_LX_0.90.npy",
    "result/clstm_lorenz/250/gc_est_LX_1.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_3.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_6.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_9.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_11.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_18.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_19.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_23.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_28.00.npy",
    "result/clstm_lorenz/250/gc_est_LX_35.00.npy",
]

fpr_all = []
tpr_all = []

# Loop through each file, calculate TPR and FPR, and store in the lists
for file_path in file_paths:
    # Load the predicted Granger causality matrix
    GC_est = np.load(file_path)
    GC_est[GC_est > 0] = 1  # Binarize the predicted matrix

    # Initialize counters
    TP = FP = TN = FN = 0

    # Loop through each element in the matrix and calculate TP, FP, TN, and FN
    for i in range(GC_true.shape[0]):
        for j in range(GC_true.shape[1]):
            if GC_est[i, j] == 1 and GC_true[i, j] == 1:
                TP += 1  # True positive
            elif GC_est[i, j] == 1 and GC_true[i, j] == 0:
                FP += 1  # False positive
            elif GC_est[i, j] == 0 and GC_true[i, j] == 0:
                TN += 1  # True negative
            elif GC_est[i, j] == 0 and GC_true[i, j] == 1:
                FN += 1  # False negative

    # Calculate TPR and FPR
    TPR = TP / (TP + FN) if (TP + FN) != 0 else 0
    FPR = FP / (FP + TN) if (FP + TN) != 0 else 0
    fpr_all.append(FPR)
    tpr_all.append(TPR)

    # Print FPR and TPR for each file
    print(f"File: {file_path}")
    print(f"TPR: {TPR}")
    print(f"FPR: {FPR}")
    print("-" * 40)

roc_auc = auc(fpr_all, tpr_all)

# Plot the ROC curve
plt.figure(figsize=(8, 6))
plt.plot(fpr_all, tpr_all, color='blue', lw=2, label=f'ROC curve (AUROC = {roc_auc:.2f})')

# Plot the diagonal reference line (performance of random guessing)
plt.plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')

# Set the limits and labels for the plot
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve')
plt.legend(loc="lower right")
plt.grid(True)

# Adjust layout and display the plot
plt.tight_layout()
plt.show()

# Output the AUROC value
print(f'AUROC = {roc_auc:.4f}')
I used different λ values to calculate TPR and FPR, but when I calculated AUROC, the result was 97, which is too different from the 81.3 ± 0.9 in the paper
![image](https://github.com/user-attachments/assets/dbde7ff0-8436-4fb6-b6ac-81cf87945856)

**Comment by weijiacheng00** [comment]
Thank you, this is very helpful to me

**Comment by weijiacheng00** [comment]
I want to know all the specific parameters of each experiment. Can you tell me

**Comment by weijiacheng00** [comment]
Do we need to change the context of each CLSTM experiment? Do we need to change the burn-in parameter for Lorenz data with different T lengths

**Comment by weijiacheng00** [comment]
if delta_t=0.1 the context is 10，and delta_t=0.05，the context is 20？

**Comment by iancovert** [comment]
Unfortunately I don't have the exact $\lambda$ values used for each experiment, I've changed institutions and no longer have access to the original source code.

If you have another method you'd like to compare with cMLP/cLSTM, I would recommend running both on the exact same dataset with the same approach to AUROC calculation. My method to find a good $\lambda$ range was to manually search for $\lambda$ values that result in full sparsity (0% variable usage) and no sparsity (100% variable usage), and then use a range of values in between. I did this separately for each dataset, which unfortunately takes some time.

For data generation under the Lorenz model, I believe we used the default `delta_t`, `sd` and `burn_in` values in our code. The current burn-in value should be large enough even if you make the sequence longer: once you've burned in, you can generate arbitrarily more data points and they'll be from the correct generating process. For `delta_t`, this shouldn't really affect the context, assuming you mean how many past data points affect the next one: my understanding is that because this is an ODE, the next point depends only on the previous point, although earlier ones may be useful to infer the actual previous point and its partial derivatives due to noise in the observations.

**Comment by weijiacheng00** [comment]
ths i try again.
